### PR TITLE
Add ONNX to TOSA lowering for onnx.Tile

### PR DESCRIPTION
**Pull request by TinaAMD**
* Lowering can only be done if the input is ranked
* The repetitions need to be constant

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/Tensor/Tile.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    int64_t inputRank = onnx_mlir::getRank(inputType);
+    auto newResultElementType = cast<ShapedType>(inputType).getElementType();
+    Type newOutputType = RankedTensorType::get(
+        llvm::SmallVector<int64_t>(inputRank, ShapedType::kDynamic),
+        newResultElementType);
+
+    // Create the attribute for the repetitions
+    Value reps = adaptor.getRepeats();
+    auto repsConstant =
+        dyn_cast_or_null<mlir::tosa::ConstOp>(reps.getDefiningOp());
```
Comment (ttjost) [inline src/Conversion/ONNXToTOSA/Tensor/Tile.cpp:57]: Could we use `if (matchPattern(reps, m_Constant()))` instead?

Comment (ttjost) [reply src/Conversion/ONNXToTOSA/Tensor/Tile.cpp:57]: Something like:
```
    DenseIntElementsAttr reps;
    if (!matchPattern(op.getRepeats(), m_Constant(&reps))) {
      return rewriter.notifyMatchFailure(op, "onnx.tile can only be lowered with constant repetitions");
    }
```

**Review by ttjost — APPROVED** [review]

**Review by mgehre-amd — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/Tensor/Tile.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace {
+
+class ONNXTileLoweringToTOSA : public OpConversionPattern<ONNXTileOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  using OpAdaptor = typename ONNXTileOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXTileOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
```
[the remark this answers is not in the document]
Comment (TinaAMD) [reply src/Conversion/ONNXToTOSA/Tensor/Tile.cpp:44]: Very good point, I added explicit checks and more testing ([bce5cfd](https://github.com/Xilinx/onnx-mlir/pull/33/commits/bce5cfdd01390269deaecf97749829b845bf2d49)). We should introduce similar checks in other lowerings, there seem to be more that do not properly check type restrictions.

File: src/Conversion/ONNXToTOSA/Tensor/Tile.cpp (hunk 1 above)
Comment (TinaAMD) [reply src/Conversion/ONNXToTOSA/Tensor/Tile.cpp]: Nice suggestion! I cleaned it up ([bce5cfd](https://github.com/Xilinx/onnx-mlir/pull/33/commits/bce5cfdd01390269deaecf97749829b845bf2d49)).

**Review by mgehre-amd — APPROVED** [review]

**Review by ttjost — APPROVED** [review]

**Pull request merged**